### DISCUSSION TITLE
In your opinion what is the best path forward if we'd like to do our own implementation of keri (with cesr/acdc) for learning or profit.

### DISCUSSION THREAD

**Discussion by daidoji** [WebOfTrust/keri#66] [Q&A]
At the point in time I'm asking this question the keri community has: 

- an excellent reference implementation in keripy
- a deployed rust keri implementation that is proprietary
- a few open source implementations of keri that were attempted but due to circumstance and economies of development were not kept current with keripy up to the present point
- 3-4 implementations of things that mostly support the signify protocols (signify-ts, signify-py, keria, etc...)

My partner and I were discussing doing another implementation of keri/acdc/cesr for both opportunity of learning, and maybe to provide inspiration or other tools for the community to use.  We have our own opinions on the best way to go about it but I thought I'd pose this question to the community as many of you have more experience with implementing keri than we do.  Namely,

- From your experience, if you were starting from day zero but knowing what you know now what's do you think the best path forward on a new implementation would be? 

ie)

- Which pieces of keri would you implement in which order?  
- Which pieces are the easiest to implement compared with some that are much more difficult?
- Would you use cesride as a base for a cesr parser (assuming we're not in js land)?
- Would you use keripy as a library shim to work toward your new implementation?
- Are there tools or test vectors you'd attempt to extract from keripy operations in order to ensure compatibility?
- How would you test compatibility with keripy (or the signify tools) in general?
- Are there any "gotchas" or areas that you'd be particularly careful with in terms of things that could be implemented quite poorly (probably the hardest question for any security bedrock type application like keri where you only need one weak link)?

We have our opinions and ideas for all of these questions but just wanted to pose it for the community (and posterity) for anyone who feels like trying out an implementation in the future.  We're just looking to pool the community resources, most of which know much more than we do before starting out.  All opinions welcome.

**Comment by kentbull** [comment]
I would implement the pieces in the following order:
1. Parser + CESR implementation in tandem
2. Async runtime event processing module for a single agent/keystore/wallet 
(this need is served by HIO + an internal queuing system in the KERIpy/KERIA impl)
4. CLI interface to parser impl and 
5. Storage Interface + Storage Implementation (LMDB or otherwise)
6. Exchange and communication protocol
7. Key Event Log Query module
8. OOBI connections
9. Challenge Phrases
10. ACDC parts - TELs, registries, IPEX methods, ACDC Schemas
11. Agent module (managing Habs)
12. Agent communication interface - REST, gRPC, etc.
13. Signify protocol

The reason you start with the parser is because that is what allows you to check your work quickly against the KERIpy reference implementation.
Then, in order to work up to the Agent module and Signify protocol you have a lot of work to do.

> Which pieces are the easiest to implement compared with some that are much more difficult?

**Easiest**
OOBI resolution, ACDC Schema handling, Challenge Phrases, ACDC Schema validation, Signify, CLI interface, Agent communication interface

**Hardest**
From what I can tell the parser, async runtime, exchange protocol, IPEX methods, and agent module are the most complex to implement since the are the core bulk of KERI. Signify is a subset of KERI that supports keys at the edge and will build on top of the core work you do.

> Would you use cesride as a base for a cesr parser (assuming we're not in js land)?

No, I would not use CESRide as a base for a CESR parser. Instead I would build a CESR code-table-aware parser that feeds byte arrays into CESR. I thought the idea was to have the parser as a base module used by the CESR implementation, as in CESRide would use Parside.

> Would you use keripy as a library shim to work toward your new implementation?

No, though it could go either way. There are so many Python idioms that it would likely make more sense to extract the mental model and start over somewhere else.

> Are there tools or test vectors you'd attempt to extract from keripy operations in order to ensure compatibility?

Yes, nearly every test in KERIpy and KERIA would need to be added in some way. Maybe not every assert, though nearly every test needs to be meaningfully represented.

> How would you test compatibility with keripy (or the signify tools) in general?

Set up an integration test using one each of a KERIA agent and your own agent implementation, or a KERIpy keystore and your own keystore impl, and then try each type of IPEX operation, delegation operation, and multisig operation where you have a mix of each implementation contributing to things like a multisig group or delegating from a KERIpy keystore to your own keystore and vice versa.

Doing this would allow you to make sure it all works well together, which is the eventual production state of things when networks grow up enough to bump into each other. 

> Are there any "gotchas" or areas that you'd be particularly careful with in terms of things that could be implemented quite poorly

First on the list is the parser, then the async runtime, followed by the database implementation, and then the Agent implementation. I remember the first agent implementation in KERIpy and I think that interface is likely the second most challenging thing to get right since it is external facing and there will be pressure to simplify, possibly oversimplify it. The parser and async runtime combo are the places where things could be implemented poorly and kill the correctness, debuggability, and performance of your app. Since it's an async system then everything internally must be non-blocking async to achieve a reasonable level of performance. You need to change your mindset to this if you are used to synchronous thinking and programming.

**Reply by daidoji** [reply]
Thanks @kentbull , appreciate it.
